compiler/rustc_transmute/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+#![feature(alloc_layout_extra, control_flow_enum)]
+#![allow(unused_imports, dead_code, unused_variables)]
+use rustc_infer::infer::InferCtxt;
+use rustc_macros::TypeFoldable;
+use rustc_middle::traits::ObligationCause;
+use rustc_middle::ty::Binder;
+use rustc_middle::ty::Ty;
+
+pub(crate) use rustc_data_structures::fx::FxHashMap as Map;
+pub(crate) use rustc_data_structures::fx::FxHashSet as Set;
+
+mod ldt;
+pub use ldt::Ldt;
+
+mod nfa;
+pub use nfa::Nfa;
+
+mod maybe_transmutable;
+pub use maybe_transmutable::maybe_transmutable;
+
+#[derive(TypeFoldable, Debug, Clone, Copy)]
+pub struct Types<'tcx> {
+    pub src: Ty<'tcx>,
+    pub dst: Ty<'tcx>,
+}
+
+/// The type encodes answers to the question: "Are these types transmutable?"
+#[derive(Debug, Hash, Eq, PartialEq, PartialOrd, Ord, Clone)]
+pub enum Answer<'tcx> {
+    /// `Src` is transmutable into `Dst`.
+    Yes,
+
+    /// `Src` is NOT transmutable into `Dst`.
+    No,
```

Comment:
Re. better diagnostics: I suspect eventually we'll make this `No` carry information about the incompatibility that makes the transmute invalid. The challenge with this is going to be ensure that the *single* counter-example that's eventually surfaced in the diagnostic is the most useful one.